synapse/api/constants.py

```diff
@@ -39,217 +41,217 @@ class Membership:
 
     """Represents the membership states of a user in a room."""
 
-    INVITE = "invite"
-    JOIN = "join"
-    KNOCK = "knock"
-    LEAVE = "leave"
-    BAN = "ban"
-    LIST = (INVITE, JOIN, KNOCK, LEAVE, BAN)
+    INVITE: Final = "invite"
+    JOIN: Final = "join"
+    KNOCK: Final = "knock"
+    LEAVE: Final = "leave"
+    BAN: Final = "ban"
+    LIST: Final = (INVITE, JOIN, KNOCK, LEAVE, BAN)
```

Comment:
> `Revealed type is "Tuple[Literal['invite']?, Literal['join']?, Literal['knock']?, Literal['leave']?, Literal['ban']?]"`

Actually, the earlier lines print things like `Revealed type is "Literal['invite']?"` too.

Do we know why it says a question mark after each Literal? ... It makes me worry that something secret is going on.

Reply:
Good question. https://mypy.readthedocs.io/en/stable/literal_types.html#declaring-literal-variables says

> If you do not provide an explicit type in the Final, the type of c becomes context-sensitive: mypy will basically try “substituting” the original assigned value whenever it’s used before performing type checking. This is why the revealed type of c is Literal[19]?: the question mark at the end reflects this context-sensitive nature.

Reply:
Thanks for the link David. It took me a few reads to see what it's going on about, but interesting — question is: do we want the behaviour given by `Final` or the one given by `Literal[…]`?

Reply:
> do we want the behaviour given by `Final` or the one given by `Literal[…]`?

I'm having trouble coming up with an example where it makes a difference. Do you have one in mind?

Reply:
It's really hard to come up with an example where it makes a difference. Here's one:
```py
T = TypeVar("T", bound=str)
def validate_or_throw(value: str, allowed_values: List[T]) -> T:
    ...

ACTIONS_IMPLICIT: Final = ["ban", "unban"]
ACTIONS_EXPLICIT: List[Literal["ban", "unban"]] = ["ban", "unban"]

def do_action(action: Literal["ban", "unban"]):
    pass

def on_GET1(json: Dict):
    action = validate_or_throw(json["action"], ACTIONS_IMPLICIT)
    do_action(action)  # Argument 1 to "do_action" has incompatible type "str"

def on_GET2(json: Dict):
    action = validate_or_throw(json["action"], ACTIONS_EXPLICIT)
    do_action(action)  # ok
```

I'm tempted to stick with `Final` until the context-sensitive nature becomes a problem.

Reply:
> I'm tempted to stick with Final until the context-sensitive nature becomes a problem.

I'm fine with this---no worse than the current situation as far as I can see

Reply:
aye yeah, let's just stick with it